enable full percentile spectrum analysis on latency

filipecosta90 opened the pull request:
# Full percentile spectrum analysis
Fixes: #68 
This PR enable full percentile spectrum analysis on query latency so that we can fully understand the performance and stability characteristics of the system we are measuring. 

Right now memtier_benchmark is displaying single values targets like the average. 
Considering that latency does not obey a normal distribution we should enable ourselves to do full percentile spectrum analysis. Just looking at a number might be misleading. 

A useful visualization for pointing out the folly of chasing a mean measurement is illustrated below - they have nearly identical simple descriptive statistics, yet have very different distributions and appear very different when graphed.

![image](https://user-images.githubusercontent.com/5832149/67484927-f4734880-f660-11e9-9400-6e0dc09a8a87.png)


Source: https://en.wikipedia.org/wiki/Anscombe%27s_quartet

_____
_______
# Enabling full percentile latencies on memtier_benchmark:
The previous features can be applied with the changes added on this PR using the following added options:
```
      --hdr-file-prefix=FILE     Prefix of HDR Latency Histogram output files, if not set, will not save latency histogram files
      --print-percentiles          Specify which percentiles info to print on the results table (by default prints percentiles: 50,99,99.9)

```

filipecosta90 commented on run_stats.cpp:
```diff
@@ -479,17 +680,26 @@ void run_stats::debug_dump(void)
+    percentiles = &iter.specifics.percentiles;
+    while (hdr_iter_next(&iter)){
+        double value = iter.highest_equivalent_value / (double) LATENCY_HDR_RESULTS_MULTIPLIER;
+        benchmark_debug_log("  SET <= %u msec: %u\n", value,percentiles->percentile);
```
I agree that it's too much of a verbose log. and if people want the hdr output they can simply add the filename parameter and have outputted there. what do you think?

yossigo replied:
I agree.

filipecosta90 commented on run_stats.cpp:
```diff
@@ -422,18 +434,207 @@ void run_stats::save_csv_arbitrary_commands(FILE *f, arbitrary_command_list& com
+        if (!hdr_outfile){
+            perror(config->hdr_prefix);
+            return false;
+        }
```
splitted into two generic methods ( one for each file type ). what do you think?

yossigo replied:
I think it looks much better now!

yaacovhazan-Redislabs commented on run_stats.h:
```diff
@@ -28,11 +28,14 @@
-typedef std::map<float, int> latency_map;
-typedef std::map<float, int>::iterator latency_map_itr;
-typedef std::map<float, int>::const_iterator latency_map_itr_const;
+#define LATENCY_HDR_MIN_VALUE 1
```
if the only usage of those #define's are in run_stats.h pls consider to move them (to run_stats.cpp)..

yaacovhazan-Redislabs commented on shard_connection.h:
```diff
@@ -172,6 +174,8 @@ class shard_connection {
     struct bufferevent *m_bev;
     struct event_base* m_event_base;
 
+    bool limited;
```
the only usage of this one is in fill_pipeline... why we can't use local variable instead of class's member?

filipecosta90 replied:
💯 agree. corrected in the latest commit.

yaacovhazan-Redislabs commented on shard_connection.cpp:
```diff
@@ -542,6 +572,23 @@ void shard_connection::send_set_command(struct timeval* sent_time, const char *k
+ * latency of the request as well.
+ * In this way we are not effecting other clients overall latency measurement.
+ */
+void shard_connection::send_empty_rate_limit_event() {
```
if I understood correctly, every time that we want to suspend the requests .. we send error command(s) to the server?
isn't the latency of the command can affect the rate limit?

filipecosta90 replied:
@yaacovhazan-Redislabs , we write data to the file descriptor without going to the server at all. We simply write `"^\r\n"` just so that it can be discarded correctly on `void shard_connection::process_response(void)`.

filipecosta90 commented on memtier_benchmark.cpp:
```diff
@@ -1156,6 +1191,16 @@ int main(int argc, char *argv[])
     config_init_defaults(&cfg);
+    // We need this here since the the defaults might change the calculations
+    // The rate of commands per second needs to be higher that the total number of connections
+    if ( ( cfg.rate_limit_rps > 0 ) && ( cfg.rate_limit_rps < cfg.clients*cfg.threads) ) {
```
given the example command:

```
memtier_benchmark --hide-histogram -n 10 --rate-limit-rps 100
```

Not forcing can lead to situations in which we have a final correct average RPS that will target the described rps, but we will have instantaneous rps of 0 and 200 rps, 0 and 200 again. So we will have bursts followed by seconds without requests in a cyclic pattern. 

```
[RUN #1 8%,  10 secs]  4 threads:        1600 ops,     199 (avg:     159) ops/sec, 12.69KB/sec (avg: 9.77KB/sec),  1.82 (avg:  2.03) msec latencyy
[RUN #1 11%,  17 secs]  4 threads:        2200 ops,       0 (avg:     129) ops/sec, 0.00KB/sec (avg: 7.92KB/sec),   nan (avg:  2.00) msec latencyy
[RUN #1 14%,  22 secs]  4 threads:        2800 ops,     199 (avg:     127) ops/sec, 11.50KB/sec (avg: 7.84KB/sec),  1.88 (avg:  1.98) msec latency
[RUN #1 100%, 194 secs]  0 threads:       20000 ops,       0 (avg:     103) ops/sec, 0.00KB/sec (avg: 6.28KB/sec),   nan (avg:  2.02) msec latency
```
Will mention it on command line usage :)

yaacovhazan-Redislabs commented on shard_connection.cpp:
```diff
@@ -542,6 +572,23 @@ void shard_connection::send_set_command(struct timeval* sent_time, const char *k
+ * In this way we are not effecting other clients overall latency measurement.
+ */
+void shard_connection::send_empty_rate_limit_event() {
+    if ( bufferevent_write(m_bev,"^\r\n",3) == 0 ){
```
Please check it again... when we use bufferevent_write we are adding data to the output buffer of m_bev ... means.. writing data on the socket to the server

filipecosta90 replied:
you're totally right @yaacovhazan-Redislabs . 
confirmed that yes, we're sending the "rate-limiting-dummy" command to redis.  will correct it 👍

yossigo replied:
@filipecosta90 The libevent `bufferevent` has some built in rate limiting mechanism. It's obviously based on throughput and knows nothing about requests/responses but perhaps it's something you can leverage for this. I don't remember how it works or what APIs are exposed so it may or may not be possible but worth looking.

filipecosta90 replied:
hi there @yossigo and @yaacovhazan-Redislabs, sorry for the delayed reply. Considering that the histogram and the rate-limiting features are separate ones and one is stalling other I would go for splitting this PR into three:
- 1st one ( this one ) -  with overall percentile analysis feature
- 2nd one - extended percentile over time ( as discussed with @yossigo on January )
- 3rd one - rate limiting feature using libevent's Bufferevents and Rate-limiting features 
In that manner we would not delay the possibility of having extended metrics on memtier. What do you guys think?

filipecosta90 commented:
Hi there @yossigo and @yaacovhazan-Redislabs I've pruned the PR and I believe it's ready for a new review.
I've updated the hdrhistogram to the latest official one ( v0.11.0 ), added a test case for the added features, and extended the PR to store per second latency/throughput metrics at the output json ( The stored percentiles comes from the `--print-percentiles` property ). Here is a snippet:
```
,"ALL STATS":{
		"Runtime":{
			"Start time": 1597016193374
			,"Finish time": 1597016207579
			,"Total duration": 14205
			,"Time unit": "MILLISECONDS"
			}
		,"Sets":{
			"Count": 182000
			,"Ops/sec": 12812.51
			,"Hits/sec": 0.00
			,"Misses/sec": 0.00
			,"Latency": 1.419
			,"Average Latency": 1.419
			,"Min Latency": 0.032
			,"Max Latency": 14.415
			,"KB/sec": 986.79
			,"Time-Serie":{
				"0":{
					"Count": 14225
					,"Average Latency": 1.29
					,"Min Latency": 0.54
					,"Max Latency": 6.43
					,"p50.00": 1.17
					,"p99.00": 3.73
					,"p99.90": 4.96
					}
				,"1":{
					"Count": 14739
					,"Average Latency": 1.23
					,"Min Latency": 0.10
					,"Max Latency": 7.78
					,"p50.00": 1.20
					,"p99.00": 2.62
					,"p99.90": 5.86
					}
(...)
```
WDYT?

yossigo commented:
@filipecosta90 Thanks a lot for this effort! It's a great improvement and definitely worth the effort and time invested in it.
